crates/bevy_ecs/src/archetype.rs

```diff
@@ -0,0 +1,540 @@
+use crate::{
+    bundle::BundleId,
+    component::{ComponentFlags, ComponentId, StorageType},
+    entity::{Entity, EntityLocation},
+    storage::{Column, SparseArray, SparseSet, SparseSetIndex, TableId},
+};
+use bevy_utils::AHasher;
+use std::{
+    collections::HashMap,
+    hash::{Hash, Hasher},
+};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ArchetypeId(usize);
+
+impl ArchetypeId {
+    #[inline]
+    pub const fn new(index: usize) -> Self {
+        ArchetypeId(index)
+    }
+
+    #[inline]
+    pub const fn empty() -> ArchetypeId {
+        ArchetypeId(0)
+    }
+
+    #[inline]
+    pub const fn resource() -> ArchetypeId {
+        ArchetypeId(1)
+    }
+
+    #[inline]
+    pub fn index(&self) -> usize {
```

Comment:
```suggestion
    pub fn index(self) -> usize {
```
`ArchetypeId` is `Copy`.